### PR TITLE
[config] Enable/disable container monitoring when starting/stopping the services.

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

#### What I did
When we ran the command `sudo config load`, `sudo config reload` or `sudo config load_minigraph`, the containers `swss, snmp, lldp, teamd, syncd, snmp, bgp, radv, pmon, dhcp_relay, telemetry` and `restapi` would be stopped and then restarted. The script `container_checker` ran by Monit will generate false alerting messages into syslog to indicate some containers were not running during such stopping and restarting process. So this PR aims to prevent Monit from generating false alarm messages.

#### How I did it
Before stopping services, we disable Monit to monitor the running status of containers. After restarting services, we enable Monit to monitor the running status of containers again.

#### How to verify it
I deliberately reduce the monitoring interval of Monit from 60 seconds to 10 seconds to ensure the alerting messages from the script `container_checker` was generated during `sudo config reload, sudo config load and sudo config load_minigraph`. After this change was added into `_stop_services(...)` and `_restart_services(...)` , I checked that the alerting messages from `container_checker` did not appear in the syslog. 

I verified this change on the device `str-a7050-acs-3`.

#### Previous command output (if the output of a command-line utility has changed)
    admin@vlab-01:~$ sudo config reload -y
    Executing stop of service telemetry...
    Warning: Stopping telemetry.service, but it can still be activated by:
      telemetry.timer
    Executing stop of service swss...
    Executing stop of service lldp...
    Executing stop of service pmon...
    Executing stop of service bgp...
    Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
    Running command: /usr/local/bin/db_migrator.py -o migrate
    Executing reset-failed of service bgp...
    Executing reset-failed of service dhcp_relay...
    Executing reset-failed of service hostname-config...
    Executing reset-failed of service interfaces-config...
    Executing reset-failed of service lldp...
    Executing reset-failed of service ntp-config...
    Executing reset-failed of service pmon...
    Executing reset-failed of service radv...
    Executing reset-failed of service rsyslog-config...
    Executing reset-failed of service snmp...
    Executing reset-failed of service swss...
    Executing reset-failed of service syncd...
    Executing reset-failed of service teamd...
    Executing reset-failed of service telemetry...
    Executing restart of service hostname-config...
    Executing restart of service interfaces-config...
    Executing restart of service ntp-config...
    Executing restart of service rsyslog-config...
    Executing restart of service swss...
    Executing restart of service bgp...
    Executing restart of service pmon...
    Executing restart of service lldp...
    Executing restart of service telemetry...
    Reloading Monit configuration ...
    Reinitializing monit daemon

#### New command output (if the output of a command-line utility has changed)
    admin@vlab-01:~$ sudo config reload -y
    Disabling container monitoring ...
    Executing stop of service telemetry...
    Warning: Stopping telemetry.service, but it can still be activated by:
      telemetry.timer
    Executing stop of service swss...
    Executing stop of service lldp...
    Executing stop of service pmon...
    Executing stop of service bgp...
    Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
    Running command: /usr/local/bin/db_migrator.py -o migrate
    Executing reset-failed of service bgp...
    Executing reset-failed of service dhcp_relay...
    Executing reset-failed of service hostname-config...
    Executing reset-failed of service interfaces-config...
    Executing reset-failed of service lldp...
    Executing reset-failed of service ntp-config...
    Executing reset-failed of service pmon...
    Executing reset-failed of service radv...
    Executing reset-failed of service rsyslog-config...
    Executing reset-failed of service snmp...
    Executing reset-failed of service swss...
    Executing reset-failed of service syncd...
    Executing reset-failed of service teamd...
    Executing reset-failed of service telemetry...
    Executing restart of service hostname-config...
    Executing restart of service interfaces-config...
    Executing restart of service ntp-config...
    Executing restart of service rsyslog-config...
    Executing restart of service swss...
    Executing restart of service bgp...
    Executing restart of service pmon...
    Executing restart of service lldp...
    Executing restart of service telemetry...
    Enabling container monitoring ...
    Reloading Monit configuration ...
    Reinitializing monit daemon

### CODE REVIEW

File: config/main.py [hunk 1]
```diff
@@ -669,6 +669,10 @@ def _get_disabled_services_list(config_db):
 
 
 def _stop_services():
+   # Unmonitor the script of 'container_checker' before stopping services
+    click.echo("Disabling the Monit ...")
```
Comment (jleveque) [inline config/main.py:673]: Suggest rephrasing: `Disabling container monitoring ...`

File: config/main.py [hunk 2]
```diff
@@ -688,6 +692,10 @@ def _restart_services():
     click.echo("Restarting SONiC target ...")
     clicommon.run_command("sudo systemctl restart sonic.target")
 
+    # Monitor the script of 'container_checker' after restarting services
+    click.echo("Enabling the Monit ...")
```
Comment (jleveque) [inline config/main.py:696]: Suggest rephrasing: `Enabling container monitoring ...`

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — CHANGES_REQUESTED** [review]
PR title is deceptive. This change doesn't completely disable Monit, it only disables the monitoring of containers. Please update accordingly.

File: config/main.py (hunk 1 above)
Comment (yozhao101) [reply config/main.py]: Updated!

File: config/main.py (hunk 2 above)
Comment (yozhao101) [reply config/main.py]: Updated!

**Comment by yozhao101** [comment]
> PR title is deceptive. This change doesn't completely disable Monit, it only disables the monitoring of containers. Please update accordingly.

Great suggestion, Updated!

File: config/main.py [hunk 3]
```diff
@@ -669,6 +669,10 @@ def _get_disabled_services_list(config_db):
 
 
 def _stop_services():
+   # Unmonitor the script of 'container_checker' before stopping services
```
Comment (jleveque) [inline config/main.py:672]: Comment is unnecessary. Code is fairly self-explanatory.

File: config/main.py [hunk 4]
```diff
@@ -688,6 +692,10 @@ def _restart_services():
     click.echo("Restarting SONiC target ...")
     clicommon.run_command("sudo systemctl restart sonic.target")
 
+    # Monitor the script of 'container_checker' after restarting services
```
Comment (jleveque) [inline config/main.py:695]: Comment is unnecessary. Code is fairly self-explanatory.

File: config/main.py [hunk 5]
```diff
@@ -688,6 +692,10 @@ def _restart_services():
...
+    click.echo("Enabling container monitoring ...")
+    clicommon.run_command("sudo monit monitor container_checker")
```
Comment (jleveque) [inline config/main.py:697]: Suggest moving this after the Monit reload. Shouldn't make a difference, but it feels like it makes more sense.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/main.py (hunk 5 above)
Comment (yozhao101) [reply config/main.py]: I checked that moving the command `sudo monit monitor container_checker` after `sudo monit reload` did not make a difference.

**Review by jleveque — APPROVED** [review]

**Review by jleveque — APPROVED** [review]

**Comment by yozhao101** [comment]
What I did
When we ran the command sudo config load, sudo config reload or sudo config load_minigraph, the containers swss, snmp, lldp, teamd, syncd, snmp, bgp, radv, pmon, dhcp_relay, telemetry and restapi would be stopped and then restarted. The script container_checker ran by Monit will generate false alerting messages into syslog to indicate some containers were not running during such stopping and restarting process. So this PR aims to prevent Monit from generating false alarm messages.

How I did it
Before stopping services, we disable Monit to monitor the running status of containers. After restarting services, we enable Monit to monitor the running status of containers again.

How to verify it
I deliberately reduce the monitoring interval of Monit from 60 seconds to 10 seconds to ensure the alerting messages from the script container_checker was generated during sudo config reload, sudo config load and sudo config load_minigraph. After this change was added into _stop_services(...) and _restart_services(...) , I checked that the alerting messages from container_checker did not appear in the syslog.

I verified this change on the device str-a7050-acs-3.

Previous command output (if the output of a command-line utility has changed)
admin@vlab-01:~$ sudo config reload -y
Executing stop of service telemetry...
Warning: Stopping telemetry.service, but it can still be activated by:
  telemetry.timer
Executing stop of service swss...
Executing stop of service lldp...
Executing stop of service pmon...
Executing stop of service bgp...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Executing reset-failed of service bgp...
Executing reset-failed of service dhcp_relay...
Executing reset-failed of service hostname-config...
Executing reset-failed of service interfaces-config...
Executing reset-failed of service lldp...
Executing reset-failed of service ntp-config...
Executing reset-failed of service pmon...
Executing reset-failed of service radv...
Executing reset-failed of service rsyslog-config...
Executing reset-failed of service snmp...
Executing reset-failed of service swss...
Executing reset-failed of service syncd...
Executing reset-failed of service teamd...
Executing reset-failed of service telemetry...
Executing restart of service hostname-config...
Executing restart of service interfaces-config...
Executing restart of service ntp-config...
Executing restart of service rsyslog-config...
Executing restart of service swss...
Executing restart of service bgp...
Executing restart of service pmon...
Executing restart of service lldp...
Executing restart of service telemetry...
Reloading Monit configuration ...
Reinitializing monit daemon
New command output (if the output of a command-line utility has changed)
admin@vlab-01:~$ sudo config reload -y
Disabling container monitoring ...
Executing stop of service telemetry...
Warning: Stopping telemetry.service, but it can still be activated by:
  telemetry.timer
Executing stop of service swss...
Executing stop of service lldp...
Executing stop of service pmon...
Executing stop of service bgp...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Executing reset-failed of service bgp...
Executing reset-failed of service dhcp_relay...
Executing reset-failed of service hostname-config...
Executing reset-failed of service interfaces-config...
Executing reset-failed of service lldp...
Executing reset-failed of service ntp-config...
Executing reset-failed of service pmon...
Executing reset-failed of service radv...
Executing reset-failed of service rsyslog-config...
Executing reset-failed of service snmp...
Executing reset-failed of service swss...
Executing reset-failed of service syncd...
Executing reset-failed of service teamd...
Executing reset-failed of service telemetry...
Executing restart of service hostname-config...
Executing restart of service interfaces-config...
Executing restart of service ntp-config...
Executing restart of service rsyslog-config...
Executing restart of service swss...
Executing restart of service bgp...
Executing restart of service pmon...
Executing restart of service lldp...
Executing restart of service telemetry...
Enabling container monitoring ...
Reloading Monit configuration ...
Reinitializing monit daemon

**Comment by yozhao101** [comment]
@jleveque  I will create a separate PR to update the submodule.

**Comment by yxieca** [comment]
@yozhao101, @jleveque this change cannot be cherry-picked to 202012 cleanly, Is there a dependency needs to be cherry-picked?

**Comment by jleveque** [comment]
This commit has been reverted from the master branch, so I am removing the "Request for 202012 branch" label.

**Pull request merged**